File: linked_list/0061_rotate_list.js
/*
Given the head of a linked list, rotate the list to the right by k places.

The number of nodes in the list is in the range [0, 500].
-100 <= Node.val <= 100
0 <= k <= 2 * 10^9
*/

// https://leetcode.com/problems/rotate-list/description/

// make a circle, find the new head (need to know length), cut the circle
// time complexity: O(n)
// space complexity: O(1)
var rotateRight = function (head, k) {
  if (!head || !head.next || k === 0) return head;
  // calculate the length
  let length = 1,
    curr = head;
  while (curr.next) {
    length++;
    curr = curr.next;
  }
  // make it circular
  curr.next = head;
  // update k
  k %= length;
  // find the new head & tail
  curr = head;
  // find the new tail: move length - k staps
  for (i = 1; i < length - k; i++) {
    curr = curr.next;
  }
  let newHead = curr.next;
  // cut the circle
  curr.next = null;
  return newHead;
};

File: linked_list/reorder_list/0061_rotate_list.py
# Given the head of a linked list, rotate the list to the right by k places.

# The number of nodes in the list is in the range [0, 500].
# -100 <= Node.val <= 100
# 0 <= k <= 2 * 10^9

# https://leetcode.com/problems/rotate-list/

class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy_head = ListNode(0, head)
        # edge case:
        if not head or not head.next or k == 0:
            return head
        # calculate the length
        curr, l = dummy_head, 0
        while curr.next:
            curr = curr.next
            l += 1
        # make it circular
        curr.next = head
        # update k
        k %= l
        # find the new tail
        new_tail = dummy_head
        for _ in range(l - k):
            new_tail = new_tail.next
        new_head = new_tail.next
        # cut the circle
        new_tail.next = None
        return new_head